File: 백준/Gold/1167. 트리의 지름/트리의 지름.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;
using pil = pair<int, ll>;

vector<vector<pil>> e;

pil dfs(int cur, int pre){
    int node = cur;
    ll d = 0;
    for(auto [nxt, nd] : e[cur]){
        if(nxt == pre) continue;
        auto [nnde, nnd] = dfs(nxt, cur);
        nnd += nd;
        if(nnd > d) d = nnd, node = nnde;
    }
    return {node, d};
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int v; cin >> v; e.resize(v+1);
    for(int i = 0; i < v; i++){
        int x, y; cin >> x;
        while(cin >> y && y != -1){
            int z; cin >> z;
            e[x].push_back({y, z});
        }
    }
    cout << dfs(dfs(1, 0).first, 0).second;
    return 0;
}

File: 백준/Gold/1167. 트리의 지름/트리의 지름.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

sys.setrecursionlimit(10**6)

def main():
    v = int(input())
    e = [[] for _ in range(v+1)]
    for _ in range(v):
        data = [*map(int, input().split())]
        for i in range(1, len(data)-1, 2): e[data[0]].append((data[i], data[i+1]))
    def dfs(cur, pre):
        node, d = cur, 0
        for nxt, nd in e[cur]:
            if nxt == pre: continue
            nnde, nnd = dfs(nxt, cur)
            nnd += nd
            if nnd > d: d, node = nnd, nnde
        return node, d
    print(dfs(dfs(1, 0)[0], 0)[1])
    return  
if __name__ == '__main__':
    main()